### DISCUSSION TITLE
Structure of mesh generator classes (such as HoneycombMeshGenerator)

### DISCUSSION THREAD

**Discussion by zacharynevin-stemcell** [Chaste/Chaste#175] [Q&A] [answered]
Hi there,

I notice in files such as `mesh/src/HoneycombMeshGenerator.cpp` that the nodes, elements, and edges are all manually computed and written to temporary `.node`, `.elem`, and `.edge` files and used to generate the `mpMesh` instance.

Is there a reason that this can't be replaced with `MutableMesh::AddNode` method and then a call to `MutableMesh::ReMesh`? The rough code outline of what I am talking about is shown below:

```c++
// mesh/src/HoneycombMeshGenerator.cpp lines 95-130

unsigned node = 0;
// add middle layer of cells
for (unsigned i=0; i<num_nodes_along_length; i++)
{
  for (unsigned j=0; j<num_nodes_along_width; j++)
  {
    /** 
      * handle ghost node indices, boundary, compute x, y etc etc
      **/

    mpMesh->AddNode(new Node<2>(node++, boundary, x, y));
  }
}
mpMesh->ReMesh();
```

The reason I am asking is because I want to create my own mesh generator class, and it would be far easier to just compute the node positions and let the `MutableMesh` instance take care of computing the connectivity, elements, faces etc, especially if I want to work in 3D.

**Comment by mirams** [comment, marked as answer]
Hi @zacharynevin-stemcell 

You are quite right, that you could construct a mesh by defining nodes. See for example https://github.com/Chaste/Chaste/blob/3c5088640d447ad5da248b74de610b12a3a89f91/mesh/test/TestMutableMeshRemesh.hpp#L386 which uses a `MutableMesh` constructor that takes a vector of nodes (and then calls `ReMesh()` itself).
```cpp
std::vector<Node<3> *> nodes;

nodes.push_back(new Node<3>(0, true,  0.0,  0.0,  0.0));
nodes.push_back(new Node<3>(1, true,  1.0,  1.0,  0.0));
nodes.push_back(new Node<3>(2, true,  1.0,  0.0,  1.0));
nodes.push_back(new Node<3>(3, true,  0.0,  1.0,  1.0));
nodes.push_back(new Node<3>(4, false, 0.5,  0.5,  0.5));

MutableMesh<3,3> mesh(nodes);
```

In theory there could be less work involved if you can pre-define where all the nodes, edges and elements are going to be - which is why we 'manually' construct some regular meshes for (e.g.) rectangular domains (there's also an advantage in pre-defining the mesh for rectangular cases in particular where you can introduce bias into finite-element solutions by having all the diagonals aligned rather than staggered). In  the `HoneycombMeshGenerator` we do have that somewhat redundant step of writing the mesh to file and loading from there just to avoid duplicating the logic on how to build a mesh in memory in both the various mesh generators as well as mesh readers I guess. But it would also be fine and perhaps simpler to use the approach above.
